### PR TITLE
Introduce screen tabs

### PR DESCRIPTION
**Pull request by natoscott**
This is a forward port (by nathans) of Hisham's original code.

### CODE REVIEW

File: ScreenManager.c [hunk 1]
```diff
@@ -120,7 +136,50 @@ static void checkRecalculation(ScreenManager* this, double* oldTime, int* sortTi
    *rescan = false;
 }
 
+static inline bool drawTab(int* y, int* x, int l, const char* name, bool cur) {
+   attrset(CRT_colors[cur ? SCREENS_CUR_BORDER : SCREENS_OTH_BORDER]);
+   mvaddch(*y, *x, '[');
+   (*x)++;
+   if (*x >= l) return false;
```
Comment (BenBE) [inline ScreenManager.c:143]: Break those if statements onto multiple lines.

File: configure.ac [hunk 2]
```diff
@@ -367,7 +367,7 @@ if test "$enable_static" = yes; then
 fi
 if test "$my_htop_platform" = "solaris"; then
    # On OmniOS /usr/include/sys/regset.h redefines ERR to 13 - \r, breaking the Enter key.
-   # Since ncruses macros use the ERR macro, we can not use another name.
+   # Since ncurses macros use the ERR macro, we can not use another name.
```
Comment (BenBE) [inline configure.ac:370]: Unrelated bugfix. Please split into its own commit.

File: darwin/Platform.c [hunk 3]
```diff
@@ -47,6 +47,16 @@ in the source distribution for its full text.
...
 
+ScreenDefaults Platform_defaultScreens[] = {
+   {
+      .name = "Main",
+      .columns = "PID USER PRIORITY NICE M_VIRT M_RESIDENT STATE PERCENT_CPU PERCENT_MEM TIME COMM",
+      .sortKey = "PERCENT_CPU",
+   },
+};
+
+const unsigned int Platform_numberOfDefaultScreens = sizeof(Platform_defaultScreens)/sizeof(ScreenDefaults);
```
Comment (BenBE) [inline darwin/Platform.c:58]: There is an `ARRAY_SIZE` macro we could use …

File: dragonflybsd/Platform.c [hunk 4]
```diff
@@ -32,9 +32,16 @@ in the source distribution for its full text.
...
+      .name = "Main",
+      .columns = "PID USER PRIORITY NICE M_VIRT M_RESIDENT STATE PERCENT_CPU PERCENT_MEM TIME COMM",
+      .sortKey = "PERCENT_CPU",
+   },
+};
 
-const ProcessField Platform_defaultFields[] = { PID, USER, PRIORITY, NICE, M_VIRT, M_RESIDENT, STATE, PERCENT_CPU, PERCENT_MEM, TIME, COMM, 0 };
-
+const unsigned int Platform_numberOfDefaultScreens = sizeof(Platform_defaultScreens)/sizeof(ScreenDefaults);
+                                                                        32,1          13%
```
Comment (BenBE) [inline dragonflybsd/Platform.c:44]: This looks strange … I don't think this syntax issue is intended?

**Comment by smalinux** [comment]
Great work as usual, @natoscott :)

Bug 1) missing mouse handling
Settings > Screens, then choose any of them using the mouse, Htop will panic.

Bug 2) htoprc & settings related (I think!)
Try to re-order any of the "Active Columns" or just click on any of them.
then exit from Htop. re-open Htop. Will duplicate the tabs that already exist.

File: ColumnsPanel.c [hunk 5]
```diff
@@ -138,35 +138,42 @@ static void ColumnsPanel_add(Panel* super, unsigned int key, Hashtable* columns)
    Panel_add(super, (Object*) ListItem_new(name, key));
 }
 
-ColumnsPanel* ColumnsPanel_new(Settings* settings) {
+void ColumnsPanel_fill(ColumnsPanel* this, ScreenSettings* ss, Hashtable* columns) {
+   Panel* super = (Panel*) this;
+   Panel_prune(super);
+   const ProcessField* fields = ss->fields;
+   for (; *fields; fields++)
```
[the remark this answers is not in the document]
Comment (natoscott) [reply ColumnsPanel.c:145]: Nope, fixed.

File: dragonflybsd/Platform.c (hunk 4 above)
Comment (natoscott) [reply dragonflybsd/Platform.c:44]: Fixed, thanks.

**Comment by natoscott** [comment]
> Bug 1) missing mouse handling
> Bug 2) htoprc & settings related (I think!)

Thanks @smalinux - these issues should be resolved now with that last git push.

**Review by smalinux — CHANGES_REQUESTED** [review]

File: Settings.c [hunk 6]
```diff
@@ -148,51 +196,107 @@ static void Settings_defaultMeters(Settings* this, unsigned int initialCpuCount)
...
+      if (id > 0 && id < LAST_PROCESSFIELD)
+         ss->flags |= Process_fields[id].flags;
+      j++;
    }
-   settings->fields[j] = NULL_PROCESSFIELD;
    String_freeArray(ids);
 }
 
+ScreenSettings* Settings_newScreen(Settings* this, const char* name, const char* line) {
+   ScreenSettings* ss = xCalloc(sizeof(ScreenSettings), 1);
```
Comment (smalinux) [inline Settings.c:265]: I think this pointer didn't free properly. I will try to debug this later, 
ASan output:
```
==195406==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 40 byte(s) in 1 object(s) allocated from:
    #0 0x7f4adc282dc6 in calloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10ddc6)
    #1 0x559095bdcb64 in xCalloc /home/smalinux/htop/XUtils.c:55
    #2 0x559095bd3546 in Settings_newScreen /home/smalinux/htop/Settings.c:266
    #3 0x559095bd38dd in Settings_defaultScreens /home/smalinux/htop/Settings.c:289
    #4 0x559095bd3b2b in Settings_defaultScreens /home/smalinux/htop/Settings.c:285
    #5 0x559095bd3b2b in Settings_read /home/smalinux/htop/Settings.c:328
    #6 0x559095bd7fbc in Settings_new /home/smalinux/htop/Settings.c:732
    #7 0x559095b9f988 in CommandLine_run /home/smalinux/htop/CommandLine.c:303
    #8 0x7f4adbdc30b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

SUMMARY: AddressSanitizer: 40 byte(s) leaked in 1 allocation(s).
```

**Review by smalinux — CHANGES_REQUESTED** [review]

File: Settings.c (hunk 6 above)
Comment (BenBE) [reply Settings.c:265]: `xCalloc` like `calloc` takes the number of elements first and the size of each element as second. Mostly a side-note, but having them swapped may lead to unnecessary trouble down the line.

That's most likely not the bug you're chasing, but you're advised to fix it while at it. ;-)

Comment (natoscott) [reply Settings.c:265]: @smalinux thanks, I have a fix now - will push shortly.

**Comment by natoscott** [comment]
Previous few commits from @smalinux address the final remaining feedback items (incl. defaulting to tabs disabled on upgrade). If there's other areas for improvement let us know please otherwise we'll go ahead and merge soon - thanks!

**Comment by hishamhm** [comment]
@natoscott Yay, I was thinking about that branch the other day and just learned about this work! Thank you so much! ❤️

(I just learned about this by clicking through from your post on the GSoC mailing list — clearly my Github notifications are a lost cause!)

File: Action.c [hunk 7]
```diff
@@ -714,4 +756,6 @@ void Action_setBindings(Htop_Action* keys) {
    keys[KEY_F(10)] = actionQuit;
    keys[KEY_F(18)] = actionExpandCollapseOrSortColumn;
    keys[KEY_RECLICK] = actionExpandOrCollapse;
+   keys[KEY_SHIFT_TAB] = actionPrevScreen;
```
Comment (smalinux) [inline Action.c:759]: This key is Ctrl+F9
See [[Function Keys]](http://bio.gsi.de/DOCS/www.coredumps.de/doc/ibm/aix/aix_hints_and_tips/3014666.htm#4)
